Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -665,46 +689,55 @@ lemma dblXYZ_of_Z_eq_zero' {P : Fin 3 → R} (hP : W'.Equation P) (hPz : P z = 0
-    W.dblXYZ P = Units.mk0 _ (X_ne_zero_of_Z_eq_zero hP hPz) ^ 2 • ![1, 1, 0] := by
-  erw [dblXYZ_of_Z_eq_zero' hP.left hPz, smul_fin3, Units.val_pow_eq_pow_val, Units.val_mk0,
-    mul_one, mul_one, mul_zero]
+    W.dblXYZ P = P x ^ 2 • ![1, 1, 0] := by
+  erw [dblXYZ_of_Z_eq_zero' hP.left hPz, smul_fin3, mul_one, mul_one, mul_zero]
 
 lemma dblXYZ_of_Y_eq' [NoZeroDivisors R] {P Q : Fin 3 → R} (hQz : Q z ≠ 0)
     (hx : P x * Q z ^ 2 = Q x * P z ^ 2) (hy : P y * Q z ^ 3 = Q y * P z ^ 3)
     (hy' : P y * Q z ^ 3 = W'.negY Q * P z ^ 3) :
-    W'.dblXYZ P = ![eval P W'.polynomialX ^ 2, eval P W'.polynomialX ^ 3, 0] := by
+    W'.dblXYZ P = ![W'.dblU P ^ 2, W'.dblU P ^ 3, 0] := by
   rw [dblXYZ, dblX_of_Y_eq hQz hx hy hy', dblY_of_Y_eq hQz hx hy hy', dblZ_of_Y_eq hQz hx hy hy']
 
-lemma dblXYZ_of_Y_eq {P Q : Fin 3 → F} (hP : W.Nonsingular P) (hPz : P z ≠ 0) (hQz : Q z ≠ 0)
-    (hx : P x * Q z ^ 2 = Q x * P z ^ 2) (hy : P y * Q z ^ 3 = Q y * P z ^ 3)
-    (hy' : P y * Q z ^ 3 = W.negY Q * P z ^ 3) : W.dblXYZ P =
-      Units.mk0 _ ((nonsingular_iff_of_Y_eq_negY (Y_eq_negY_of_Y_eq hQz hx hy hy') hPz).mp hP).right
-        • ![1, 1, 0] := by
-  erw [dblXYZ_of_Y_eq' hQz hx hy hy', smul_fin3, Units.val_mk0, mul_one, mul_one, mul_zero]
+lemma dblXYZ_of_Y_eq {P Q : Fin 3 → F} (hQz : Q z ≠ 0) (hx : P x * Q z ^ 2 = Q x * P z ^ 2)
+    (hy : P y * Q z ^ 3 = Q y * P z ^ 3) (hy' : P y * Q z ^ 3 = W.negY Q * P z ^ 3) :
+    W.dblXYZ P = W.dblU P • ![1, 1, 0] := by
+  erw [dblXYZ_of_Y_eq' hQz hx hy hy', smul_fin3, mul_one, mul_one, mul_zero]
 
 lemma dblXYZ_of_Z_ne_zero {P Q : Fin 3 → F} (hP : W.Equation P) (hQ : W.Equation Q) (hPz : P z ≠ 0)
     (hQz : Q z ≠ 0) (hx : P x * Q z ^ 2 = Q x * P z ^ 2) (hy : P y * Q z ^ 3 ≠ W.negY Q * P z ^ 3) :
-    W.dblXYZ P = Units.mk0 _ (dblZ_ne_zero_of_Y_ne' hP hQ hPz hx hy) •
+    W.dblXYZ P = W.dblZ P •
       ![W.toAffine.addX (P x / P z ^ 2) (Q x / Q z ^ 2)
           (W.toAffine.slope (P x / P z ^ 2) (Q x / Q z ^ 2) (P y / P z ^ 3) (Q y / Q z ^ 3)),
         W.toAffine.addY (P x / P z ^ 2) (Q x / Q z ^ 2) (P y / P z ^ 3)
           (W.toAffine.slope (P x / P z ^ 2) (Q x / Q z ^ 2) (P y / P z ^ 3) (Q y / Q z ^ 3)),
         1] := by
-  have hZ {n : ℕ} : W.dblZ P ^ n ≠ 0 := pow_ne_zero n <| dblZ_ne_zero_of_Y_ne' hP hQ hPz hx hy
-  erw [dblXYZ, smul_fin3, Units.val_mk0, ← dblX_of_Z_ne_zero hP hQ hPz hQz hx hy,
-    mul_div_cancel₀ _ hZ, ← dblY_of_Z_ne_zero hP hQ hPz hQz hx hy, mul_div_cancel₀ _ hZ, mul_one]
+  have hZ {n : ℕ} : IsUnit <| W.dblZ P ^ n := (isUnit_dblZ_of_Y_ne' hP hQ hPz hx hy).pow n
+  erw [dblXYZ, smul_fin3, ← dblX_of_Z_ne_zero hP hQ hPz hQz hx hy, hZ.mul_div_cancel,
+    ← dblY_of_Z_ne_zero hP hQ hPz hQz hx hy, hZ.mul_div_cancel, mul_one]
 
 end Doubling
 
 section Addition
 
 /-! ### Addition formulae -/
 
+/-- The unit `u` such that `W.add P Q = u • ![1, 1, 0]` where `P ≠ W.neg P`. -/
+def addU (P Q : Fin 3 → F) : F :=
+  -((P y * Q z ^ 3 - Q y * P z ^ 3) / (P z * Q z))
```

Comment:
You are missing the condition `P x * Q z ^ 2 = Q x * P z ^ 2` in the docstring.

Reply:
Thanks.